relayer/middleware/metrics.middleware.ts

```diff
@@ -1,45 +1,99 @@
+import { Histogram, register } from "prom-client";
 import { Middleware } from "../compose.middleware";
-import { Context } from "../context";
 import { StorageContext } from "../storage/storage";
-import { Job } from "bullmq";
 import { Counter, Registry } from "prom-client";
 
-interface MetricsOpts {
+
+type MetricRecord = Record<string, string | number>;
+
+export interface MetricsOpts<C extends StorageContext> {
   registry?: Registry;
+  labels?: MetricLabelsOpts<C>;
+  processingTimeBuckets?: number[];
 }
 
-export function metrics(
-  opts: MetricsOpts = {},
-): Middleware<Context & { job?: Job }> {
+export class MetricLabelsOpts<C extends StorageContext> {
+  labelNames: string[];
+  customizer: (context: C) => Promise<MetricRecord>;
+
+  constructor(labelNames: string[] = [], customizer: (ctx: C) => Promise<MetricRecord>) {
+    this.labelNames = labelNames;
+    this.customizer = customizer;
+  }
+}
+
+export function metrics<C extends StorageContext>(
+  opts: MetricsOpts<C> = { },
+): Middleware<C> {
   opts.registry = opts.registry || register;
+
+
+  const getLabels = async (ctx: C, failed: boolean) => {
+    const defaultLabels = { status: failed ? "failed" : "succeeded" };
+
+    try {
+      const labels: MetricRecord = await (opts.labels?.customizer(ctx) ?? Promise.resolve({}));
+
+      const allowedKeys = opts.labels?.labelNames ?? [];
+      const validatedLabels: typeof labels = {};
+      allowedKeys.forEach(key => {
+        if (labels[key]) {
+          validatedLabels[key] = labels[key];
+        }
+      });
+
+      return { ...validatedLabels, ...defaultLabels };
+    } catch (error) {
+      ctx.logger?.error("Failed to customize metric labels, please review", error);
+
+      return defaultLabels;
+    }
+  }
+
   const processedVaasTotal = new Counter({
     name: "vaas_processed_total",
-    help: "Number of vaas processed successfully or unsuccessfully.",
+    help: "Number of incoming vaas processed successfully or unsuccessfully.",
+    labelNames: opts.labels?.labelNames ?? [],
     registers: [opts.registry],
   });
 
   const finishedVaasTotal = new Counter({
     name: "vaas_finished_total",
     help: "Number of vaas processed successfully or unsuccessfully.",
-    labelNames: ["status"],
+    labelNames: ["status"].concat(opts.labels?.labelNames ?? []),
     registers: [opts.registry],
   });
 
-  return async (ctx: StorageContext, next) => {
-    const job = ctx.storage?.job;
-    // disable this metric if storage is enabled because the storage will actually compute the metrics.
-    if (job) {
-      await next();
-      return;
-    }
+  const processingDuration = new Histogram({
+    name: `vaas_processing_duration`,
+    help: "Processing time in ms for jobs",
+    buckets: opts.processingTimeBuckets ?? [6000, 7000, 7500, 8000, 8500, 9000, 10000, 12000],
+    labelNames: ["status"].concat(opts.labels?.labelNames ?? []),
+    registers: [opts.registry],
+  });
 
+  return async (ctx: C, next) => {
```

Comment:
In general it's a good idea to add a name to this function so that it'is shown in the stack in the event of an error (otherwise in the stack is shown as `anonymous`)